🐛 Fixed overly permissive publication locale setting

betschki opened the pull request:
ref https://github.com/TryGhost/Ghost/pull/24673 − since the old PR was quite outdated, it was easier to close that and start fresh. Most changes have been migrated here, though the `LOCALE_DATA` has been moved to a static JSON due to CommonJS export/import issues.

- Added SelectWithOther component to admin-x-design-system that provides a dropdown with an "Other..." option that switches to a text input
- Added locale validation utility with support for BCP 47 language tags
- Updated PublicationLanguage to use SelectWithOther for better UX
- Added LOCALE_DATA to @tryghost/i18n with human-readable labels
- Created ESM entry point for i18n to support admin-x apps imports
- Added comprehensive unit and E2E tests

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enhances language selection UX and centralizes locale data.
> 
> - Adds `SelectWithOther` to design system (dropdown with “Other…” text input, validation, search) plus Storybook stories and exports
> - Updates `PublicationLanguage` to use `SelectWithOther`, loads options from `@tryghost/i18n/lib/locale-data.json`, and validates via new `validateLocale` utility
> - Refactors `@tryghost/i18n` to source locales from static `lib/locale-data.json`; updates CommonJS exports (`default`, `LOCALE_DATA`, `SUPPORTED_LOCALES`, `generateResources`)
> - Adds unit tests for `validateLocale`, acceptance/E2E tests for language selection, and page object updates
> - Dockerfile: ensure `ghost/i18n` is included in relevant build stages to support new imports
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e6bea5247b8b5608a3fec49dfd6e75650c1e8639. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

betschki committed:
✨ Added SelectWithOther component for publication language settings

ref https://github.com/TryGhost/Ghost/pull/24673

- Added SelectWithOther component to admin-x-design-system that provides
  a dropdown with an "Other..." option that switches to a text input
- Added locale validation utility with support for BCP 47 language tags
- Updated PublicationLanguage to use SelectWithOther for better UX
- Added LOCALE_DATA to @tryghost/i18n with human-readable labels
- Created ESM entry point for i18n to support admin-x apps imports
- Added comprehensive unit and E2E tests

betschki committed:
Fixed review comments for SelectWithOther component

- Fixed locale-validation indexOf bugs for case-insensitive 'x' detection
- Fixed stale isInitiallyCustomValue using useMemo for reactivity
- Renamed Error story to ErrorState to avoid shadowing global
- Standardized validation error messages to 'Enter a value'

betschki committed:
Fixed additional review comments

- Reject private use tag 'x' without subtags
- Enforce locale format validation on save
- Fix race condition in E2E dropdown option check

sagzy commented on apps/admin-x-settings/src/utils/locale-validation.ts:
```diff
@@ -0,0 +1,195 @@
+    // Reject if any segment is empty
+    if (trimmedValue.includes('--') || trimmedValue.startsWith('-') || trimmedValue.endsWith('-')) {
+        return errorMessage;
+    }
```
nitpick: this check should be already covered by `if (!/^[a-z0-9]+(-[a-z0-9]+)*$/i.test(trimmedValue))` on line 74, so can be deleted

sagzy commented on apps/admin-x-settings/src/utils/locale-validation.ts:
```diff
@@ -0,0 +1,195 @@
+        'zh-min', // Min Chinese (broad group)
+        'zh-min-nan', // Southern Min Chinese (Taiwanese/Hokkien) - now use "nan"
+        'zh-xiang' // Xiang/Hunanese Chinese - now use "hsn"
+    ];
```
nitpick: could define the grandfathered list at module level instead of creating it each time + use a Set for faster reads:

```
  const GRANDFATHERED = new Set([
      'en-gb-oed',
      'i-ami',
      // ...
      'zh-xiang'
  ]);

  export const validateLocale = (value: string): string | null => {
      // ...
 
     if (GRANDFATHERED(trimmedValue.toLowerCase())) {
          return null;
      }
     
     // ... 
  }
```

ibalosh commented on e2e/tests/admin/settings/publication-language.test.ts:
```diff
@@ -18,7 +18,8 @@ test.describe('Ghost Admin - i18n Newsletter', () => {
         await adminPublicationPage.goto();
         await adminPublicationPage.setLanguage('fr');
-        await expect(adminPublicationPage.languageField).toHaveValue('fr');
+        // Verify French is selected in the dropdown
```
`nit` here too, comment is not needed

betschki committed:
Refactored locale validation and cleaned up comments

- Moved grandfathered tags to module-level Set for faster lookup
- Removed self-explanatory comments
